runtime-modules/project-token/src/lib.rs

```diff
@@ -0,0 +1,479 @@
+        map
+            hasher(blake2_128_concat) T::Hash => ();
+    }
+
+}
+
+decl_module! {
+    /// _MultiCurrency_ substrate module.
+    pub struct Module<T: Trait> for enum Call
+    where
+        origin: T::Origin
+    {
+
+        /// Default deposit_event() handler
+        fn deposit_event() = default;
+
+        /// Predefined errors.
+        type Error = Error<T>;
+
+        /// Transfer `amount` from `src` account to `dst` according to provided policy
+        /// Preconditions:
+        /// - `token_id` must exists
+        /// - `dst` underlying account must be valid for `token_id`
+        /// - `src` must be valid for `token_id`
+        /// - `dst` is compatible con `token_id` transfer policy
+        ///
+        /// Postconditions:
+        /// - `src` free balance decreased by `amount` or removed if final balance < existential deposit
+        /// - `dst` free balance increased by `amount`
+        /// - `token_id` issuance eventually decreased by dust amount in case of src removalp
+        #[weight = 10_000_000] // TODO: adjust weight
+        pub fn transfer(
+            origin,
+            token_id: T::TokenId,
+            outputs: TransfersOf<T>,
+        ) -> DispatchResult {
+            let src = ensure_signed(origin)?;
+
+            // Currency transfer preconditions
+            Self::ensure_can_transfer(token_id, &src, &outputs)?;
+
+            // == MUTATION SAFE ==
+
+            Self::do_transfer(token_id, &src, &outputs);
+
+            Self::deposit_event(RawEvent::TokenAmountTransferred(
+                token_id,
+                src,
+                outputs.into(),
+            ));
+            Ok(())
+        }
+
+        #[weight = 10_000_000] // TODO: adjust weight
+        pub fn dust_account(origin, token_id: T::TokenId, account_id: T::AccountId) -> DispatchResult {
+            let sender = ensure_signed(origin)?;
+            let token_info = Self::ensure_token_exists(token_id)?;
+            let account_to_remove_info = Self::ensure_account_data_exists(token_id, &account_id)?;
+
+            Self::ensure_user_can_dust_account(
+                &token_info.transfer_policy,
+                &sender,
+                &account_id,
+                &account_to_remove_info,
+            )?;
+
+            // == MUTATION SAFE ==
+
+            AccountInfoByTokenAndAccount::<T>::remove(token_id, &account_id);
+
+            let bloat_bond = T::BloatBond::get();
+            let _ = T::ReserveCurrency::deposit_creating(&account_id, bloat_bond);
+
+            Self::deposit_event(RawEvent::AccountDustedBy(token_id, account_id, sender, token_info.transfer_policy));
+
+            Ok(())
+        }
+
+        /// Join whitelist for permissioned case: used to add accounts for token
+        /// Preconditions:
+        /// - 'token_id' must be valid
+        /// - transfer policy is permissionless or transfer policy is permissioned and merkle proof is valid
```

Comment:
```suggestion
        /// - transfer policy is permissioned and merkle proof is valid
```
The method throws if transfer policy is permissionless

Reply:
:+1: